### ISSUE TITLE
microsoft authflow

### ISSUE THREAD

**Issue by dkwo** [pdobsan/oama#41]
it seems thunderbird recently removed their client_secret for microsoft, and switched to another method.
at least, I'm unable to get a new token using their older client_id and secret within mutt/oauth2.py
(this was triggered for me by a password change. it may have been broken for a while.)

I understand this question is not specific to mailctl/oama, but does anyone know a working way to get a token for microsoft accounts, in case the user cannot register an app themselves, like with universities or corporations?

in other words, how do you satisfy the requirement 'You need to provide your own client_id and client_secret of your application or of a suitable FOSS registered application.' for this specific case?

I'm out of ideas. Thanks.

edit: thunderbird ref https://bugzilla.mozilla.org/show_bug.cgi?id=1685414

**Comment by mcrosson** [comment]
I had to do the following to get things mostly working with o365. Formatted as a rough list from my notes before I gave up on using oama for the time being. I ended up using the `mutt_oauth.py` script with gpg and my ids/secrets set to get a valid xoauth2 token for use with offlineimap.

- My current blocker: microsoft now requires https as the redirect url in the app config on their side *and* that url has to match anything done on the oama side. From what I can sort oama doesn't do ssl on the embededed web server which is a problem and seems to block oama from working with current o365 requirements.
- I managed to figure out how to create an app+credential that 'works' but I own my o365 account and have full admin rights. My approach may *not* work for 'crossing tenants'.
- new app creation via o365 admin portal
  - admin centers -> all admin centers
  - microsoft entra
  - identity -> applications -> enterprise applications
  - new application
  - pick the non-gallery option and give it a name
  - overview has app id which maps to the client_id in oama config
  - properties -> assignment required & visible to users set to 'No'
  - properties -> find 'additional properties' note and click 'application registration link'
  - authentication -> supported account types set to multi tenant
  - authentication -> add platform, pick web, enter redirect uri (requires https, this is a blocker with oama)
  - certificates & secrets -> new client secret (use 24 months, thats the limit ms imposes)
  - api permissions -> add -> microsoft graph -> delegated permissions
     - open id perms (all)
     - IMAP.AccessAsUser.All
     - others as appropriate for use case

**Comment by dkwo** [comment]
@mcrosson Thanks. Unfortunately, I don't think I have admin access in my organization. What worked for me, was to use the new thunderbird `client_id` with an empty secret in `mutt/oauth2.py` script.

**Comment by Notmarrco** [comment]
thank you @dkwo , it works indeed with thunderbird client_id and `mutt_oauth2.py`.

I haven't tried but gnome evolution also has a public client_id that can be used: https://gitlab.gnome.org/GNOME/evolution/-/wikis/EWS-OAuth2

@pdobsan I would like to use oama instead of mutt_oauth2.py, so that I can have all my conf and secrets managed by a single tool.
IIUC using an already existing client_id stops us from using a "localhost" redirect_uri, since thunderbird or evolution were not configured to do that. (and I don't know of any "open source" client_id having a localhost redirect_uri)

Could it be possible to add at least "devicecode"  authentication method in oama please ? it needs user action in copying-pasting the code in the browser, so it's less "automatic" than oama's actual workflow, but it could allow us to use thunderbird client_id in oama, since microsoft client creation is so cumbersome.
thanks anyway for the software !

**Comment by pdobsan** [comment]
On Tue, Dec 03, 2024 at 08:24:28AM -0800, notmarrco wrote:
> IIUC using an already existing client_id stops us from using a
> "localhost" redirect_uri, since thunderbird or evolution were not
> configured to do that. (and I don't know of any "open source"
> client_id having a localhost redirect_uri)

The redirect_uri has nothing to do with the client_id whatsoever. oama
works with any valid client_id/client_secret pair regardless they belong
to a registered email client or were individually generated/registered
by a user.

**Comment by Notmarrco** [comment]
I may be wrong, but from the bugzilla link given by @dkwo , they're saying that there is (at least for microsoft) 2 ways of registering a client : with the client_id/client_secret pair, and the other one as a "public client".
The public client doesn't have a client_secret, and must use another flow (like "devicecode"), asking delegation from the user. https://learn.microsoft.com/en-us/entra/identity-platform/v2-app-types#mobile-and-native-apps . 

The public client is typically for email apps like thunderbird, and secret client are typically other web services where the client_secret can stay _secret_.

My comprehension is that with client_secret, oama doesn't have to _prove_ that it's the one doing the token request, so the users just have to accept connecting oama to their account. When using a public client, there must be a confirmation from the user that the application asking the rights is the right one, by confirming for example that the same code appears on oama and on the microsoft page asking the confirmation.

**Comment by pdobsan** [comment]
solved with --device option
